lib/privileges.ts

```diff
@@ -0,0 +1,272 @@
+  try {
+    checkPrivileges(ability, [{ action: "view", subject: "Privilege" }]);
+    return await prisma.privilege.findMany({
+      select: {
+        id: true,
+        nameEn: true,
+        nameFr: true,
+        descriptionEn: true,
+        descriptionFr: true,
+        permissions: true,
+        priority: true,
+      },
+      orderBy: {
+        priority: "asc",
+      },
+    });
+  } catch (e) {
+    return prismaErrors(e, []);
+  }
+};
+
+export const updatePrivilege = async (ability: Ability, privilege: Privilege) => {
+  try {
+    checkPrivileges(ability, [{ action: "update", subject: "Privilege" }]);
+
+    const response = await prisma.privilege.update({
+      where: {
+        id: privilege.id,
+      },
+      data: privilege,
+
+      select: {
+        id: true,
+      },
+    });
+    // Flush existing privilege cache for all users asynchronously
+    flushValues();
+    return response;
+  } catch (error) {
+    logMessage.error(error as Error);
+    if (error instanceof Prisma.PrismaClientKnownRequestError && error.code === "P2025") {
+      // Error P2025: Record to update not found.
+      return null;
+    }
+    throw error;
+  }
+};
+
+export const createPrivilege = async (ability: Ability, privilege: Privilege) => {
+  try {
+    checkPrivileges(ability, [{ action: "create", subject: "Privilege" }]);
+
+    const response = await prisma.privilege.create({
+      data: privilege,
+
+      select: {
+        id: true,
+      },
+    });
+    // Flush existing privilege cache for all users asynchronously
+    flushValues();
+    return response;
+  } catch (error) {
+    logMessage.error(error as Error);
+    if (error instanceof Prisma.PrismaClientKnownRequestError && error.code === "P2025") {
+      // Error P2025: Record to update not found.
+      return null;
+    }
+    throw error;
+  }
+};
+
+/**
+ * Checks the privileges requested against an ability instance and throws and error if the action is not permitted.
+ * @param ability The ability instance associated to a User
+ * @param rules An array of rules to verify
+ * @param logic Use an AND or OR logic comparison
+ */
+export const checkPrivileges = (
+  ability: Ability,
+  rules: {
+    action: Action;
+    subject: Subject | ForcedSubjectType;
+  }[],
+  logic: "all" | "one" = "all"
+): void => {
+  // helper to define if we are force typing a passed object
+  function isForceTyping(subject: Subject | ForcedSubjectType): subject is ForcedSubjectType {
+    return (
+      (subject as ForcedSubjectType).type !== undefined &&
+      (subject as ForcedSubjectType).object !== undefined
+    );
+  }
+  const result = rules.map(({ action, subject }) => {
+    if (isForceTyping(subject)) {
+      return ability.can(action, setSubjectType(subject.type, subject.object));
+    } else {
+      return ability.can(action, subject);
+    }
+  });
+  let accessAllowed = false;
+  if (process.env.NODE_ENV !== "production") {
```

Comment:
Pino does have the checking in place to only log the debug level in not prod and info level in prod.  What this environment check if for is to avoid the app going through the map function in prod only to not log the debug strings.  Slight optimization for the production environment only.